Add: ks-nav tool set

alessandrocarminati opened the pull request:
ks-nav is a set of two tools developed to support ELISA activities.

Signed-off-by: Alessandro Carminati <acarmina@redhat.com>
Signed-off-by: Maurizio Papini <mpapini@redhat.com>

alessandrocarminati committed:
Add: ks-nav tool set

ks-nav is a set of two tools developed to support ELISA activities.

Signed-off-by: Alessandro Carminati <alessandro.carminati@gmail.com>
Signed-off-by: Alessandro Carminati <acarmina@redhat.com>
Signed-off-by: Maurizio Papini <mpapini@redhat.com>

alessandrocarminati committed:
Bulk update

* Functionality add: Labels to arcs in nav subsystems mode
* Functionality add: targeted view mode
* Functionality add: references to the code for indirect calls
* Functionality add: switch to indicate "stop at" or "stop before"
* Fix: Replace "Undefined" with "THE REST" following the MAINTAINERS file
  nomenclature
* Fix: the bug to consistently spot indirect calls on x86
* Minor code refactor

The display tool nav [1] currently supports two modes of operation:
display graph at the function level, where nodes represent functions
display graph at the subsystem level , where nodes represent subsystems
in this latter mode, for a detailed analysis, it is useful to have a label
over each arch representing the symbols that contribute to the arch.
This task requires adding this functionality.

Graphs generated by the tool[1] are complex, and in certain situation, it
could be useful to have a view where only interesting subsystems are shown.
This task aims to add this feature.

During The ELISA meeting on Dec 6th, 2022 its been requested to replace the
"undefined" subsystem with "The REST".
Subsystems names are taken from the MAINTAINERS file. All the subsystems
are defined through a list of directories and files. All the files that
do not apply any subsystem are defined as "the rest".
The tool currently uses "undefined", but during the meeting, it's been
requested to name this subsystem "THE REST" as it appears in the
MAINTAINERS file.

Indirect calls are all these calls that use a variable pointer in place of
a call to a literal one.
Usually, all indirect calls are implemented at a low level as call
instruction (BL, CALL, JAL) that uses registers as an argument.
All this works like a charm until you take into consideration the
speculative execution bugs like Spectre.
To mitigate the vulnerability, the intel version of the Linux kernel
replaced a consistent number of indirect calls using registers with
retptoline[2].
Because of this, the strategy of finding indirect call instruction does not
work well on kernels compiled for intel.

When an indirect call is spotted, it is useful to have a reference to the
point in the source code where this happens.
At the binary level, the instruction sequence implementing the indirect
call is meaningless, but looking at the same source code, variable names
and context can be a valuable suggestion for the operator to resolve the
call.

The navigation tool performs an exploration starting from a given symbol
name.
In certain cases, it is useful for this exploration to exclude some
branches, because, for example, are not of interest.
In this context, there are two interpretations of this branch exclusion:

* explore the call tree and when arrived at an excluded symbol, insert it
  in the tree and proceed no further.
* explore the call tree and when arrived at an excluded symbol do not
  proceed and do not insert the symbol in the resulting graph.

[1]. https://github.com/alessandrocarminati/nav
[2]. https://lkml.org/lkml/2018/1/3/780

Signed-off-by: Alessandro Carminati <alessandro.carminati@gmail.com>
Signed-off-by: Alessandro Carminati <acarmina@redhat.com>
Signed-off-by: Maurizio Papini <mpapini@redhat.com>
Signed-off-by: Veronika Fuxova <vfuxova@redhat.com>

reiterative commented on ks-nav/CONTRIBUTING.md:
```diff
@@ -0,0 +1,19 @@
+
+The following is a set of guidelines for contributing to nav.
+The nav utility is hosted in the 
+[nav repository](https://github.com/alessandrocarminati/nav) 
```
Should this refer to the ELISA GitHub repo instead?

alessandrocarminati replied:
Currently, I am developing the tool, which is not yet finished in the repository that is listed in the README file. The link's intent was to provide the source where a user could find the newest version. If required, I can change them instantly.

paolonig replied:
I think we should move everything to the ELISA github and remove personal repos

paolonig replied:
This has been solved in pull request 12

reiterative commented on ks-nav/kern_bin_db/CONTRIBUTING.md:
```diff
@@ -0,0 +1,19 @@
+
+The following is a set of guidelines for contributing to kern_bin_db.
+The kern_bin_db utility is hosted in the 
+[kern_bin_db repository](https://github.com/alessandrocarminati/kern_bin_db-) 
```
Should this refer to the ELISA GitHub repo instead?

alessandrocarminati replied:
Same as above.

paolonig replied:
Same as above

paolonig replied:
This has been solved in pull request 12

reiterative commented:
Are the MAINTAINERS files intended to be included in this repository, or should they be provided by the user as input for a specific analysis?

alessandrocarminati commented:
@reiterative  The "MAINTAINERS" file is specific to the kernel version under analysis. It is a living object that changes accordingly with changes in the kernel code base. 
The tool that analyse the kernel image is supposed to access to the proper "MAINTAINERS" file during the analysis to be able to provide coherent data.

alessandrocarminati committed:
Refactor code, abstract persistence operations from SQL, and improve tests

This commit includes several changes to the codebase:

- Persistence operations have been abstracted from the SQL code, making it
  easier to switch between different data storage systems.
- Tests have been improved to provide better coverage and reliability.
- Source code references have been added to some of the diagrams to aid in
  understanding.
- Graphviz image generation has been incorporated, making it easier to
  visualize the diagrams.
- Support for multiple databases has been added, allowing for greater
  flexibility in deployment.
- Several bugs have been fixed, improving the overall stability and
  functionality of the application.

Signed-off-by: Alessandro Carminati <alessandro.carminati@gmail.com>
Signed-off-by: Maurizio Papini <maurizio.papini@gmail.com>
Signed-off-by: Lev Veyde <lveyde@gmail.com>
Signed-off-by: Leonardo Rossetti <lrossett@redhat.com>
Signed-off-by: Veronika Fuxova <vfuxova@redhat.com>
Signed-off-by: Rob Linden <rokkbert@gmail.com>

paolonig commented:
This pull request is superseded by pull #12
